File: common-crypto/src/main/java/com/arsvechkarev/commoncrypto/AesSivTinkCipher.java
package com.arsvechkarev.commoncrypto;

import com.google.crypto.tink.subtle.AesSiv;
import java.nio.charset.StandardCharsets;
import java.security.GeneralSecurityException;
import java.security.MessageDigest;

public class AesSivTinkCipher {

  public static byte[] encrypt(String masterPassword, String plaintext)
      throws GeneralSecurityException {
    byte[] hashedPassword = MessageDigest.getInstance("SHA-512")
        .digest(masterPassword.getBytes(StandardCharsets.UTF_8));
    return new AesSiv(hashedPassword)
        .encryptDeterministically(plaintext.getBytes(StandardCharsets.UTF_8), null);
  }

  public static String decrypt(String masterPassword, byte[] ciphertext)
      throws GeneralSecurityException {
    byte[] hashedPassword = MessageDigest.getInstance("SHA-512")
        .digest(masterPassword.getBytes(StandardCharsets.UTF_8));
    return new String(new AesSiv(hashedPassword)
        .decryptDeterministically(ciphertext, null), StandardCharsets.UTF_8);
  }
}

File: common-crypto/src/main/java/com/arsvechkarev/commoncrypto/AesSivTinkCipher.kt
package com.arsvechkarev.commoncrypto

import com.google.crypto.tink.subtle.AesSiv
import java.nio.charset.StandardCharsets
import java.security.MessageDigest

object AesSivTinkCipher {
  
  fun encrypt(masterPassword: String, plaintext: String): ByteArray {
    val hashedPassword = MessageDigest.getInstance("SHA-512")
        .digest(masterPassword.toByteArray(StandardCharsets.UTF_8))
    val daead = AesSiv(hashedPassword)
    return daead.encryptDeterministically(
      plaintext.toByteArray(StandardCharsets.UTF_8), null)
  }
  
  fun decrypt(masterPassword: String, ciphertext: ByteArray): String {
    val hashedPassword = MessageDigest.getInstance("SHA-512")
        .digest(masterPassword.toByteArray(StandardCharsets.UTF_8))
    val daead = AesSiv(hashedPassword)
    return String(daead.decryptDeterministically(ciphertext, null),
      StandardCharsets.UTF_8)
  }
}
